core/trino-main/src/main/java/io/trino/operator/output/BlockTypeDispatchingPositionsAppender.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.output;
+
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilderStatus;
+import io.trino.spi.block.DictionaryBlock;
+import io.trino.spi.block.RunLengthEncodedBlock;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+
+import static java.util.Objects.requireNonNull;
+
+public class BlockTypeDispatchingPositionsAppender
+        implements BlockTypeAwarePositionsAppender
+{
```

Comment:
Let's skip `BlockTypeDispatchingPositionsAppender` for now. I don't think there is a case where multiple levels of nesting is neccecery (please prove me wrong). Ideally there is just:
```
RLE -> type block
dictionary -> type block
```

Single level of nesting would be unnested via `DictionaryAwarePositionsAppender/AdaptivePositionsAppender`

cc @martint